Renesas R-Car: Add Pinmux driver version 2

julien-massot opened the pull request:
Hi,
This pull request supersedes #34759.

Introduce driver for the Pin Function Controller, that can be found on various SoCs
of the R-Car Gen3 serie.
This driver allows to set a pin as GPIO or Peripherals, to set a peripheral function,
and to set bias as well as driving capabilities to a pin.

Each pins have to be declared in a dt-binding include file,  as well as the pin function parameters (see pinctrl-r8a77951.h).
Bias and drives registers content have to be describe in a per soc file (see pfc_r8a77951.c).

The dts binding differs from the Linux one that can be found here:
https://elixir.bootlin.com/linux/v5.14.12/source/Documentation/devicetree/bindings/pinctrl/renesas,pfc.yaml
- Pins are identified with uint16_t type instead of strings to avoid having to store a mapping of string to uint16_t for each pin
of the SoCs, so that only needed information for a pin are built into the Zephyr firmware.
- Function are also describes using macro for the same reason.
- We do not support pin group at the moment.
- We do not support grandchildren pin definition
- Bias Pull-up and Drive strength have the same binding and the same implementation,
the different register description are stored in struct arrays, and the corresponding registers, bit position,
or cell size for these parameters are found at runtime.

On R-Car some pins can be used as GPIO, that's the one defined with RCAR_GP_PIN(bank, pin),
macro, some can only acts as peripheral and are defined using RCAR_NOGP_PIN(pin).

Commit:
dts: bindings: pinctrl: add Renesas R-Car Pinmux

Renesas R-Car series contains a PFC (Pin Function Controller).
This module consists of registers for selecting the function of
the multiplexed pins and controls the pull-up resistor on each pin.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

Commit:
dts: arm: r8a77951: add pin-controller node

Pin controller address is the same for all member of the
GEN3 SoC Series, but pinmux configuration is SoC specific.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

Commit:
drivers: can: RCar add pinctrl support

Get pin configuration from device tree to
set alternate pin function.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

gmarull commented:
Thanks for this PR, I'll try to review it in the upcoming days. Would you mind providing a link to the reference manual?

> * We do not support pin group at the moment.
> * We do not support grandchildren pin definition

Note that this is a design choice, can't be added later. There's now one driver using grouping model that may be used as a reference in case you are interested: https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/pinctrl/gd%2Cgd32-pinctrl-af.yaml

julien-massot commented:
> Thanks for this PR, I'll try to review it in the upcoming days. Would you mind providing a link to the reference manual?
There is one named "R-Car Series, 3rd Generation User’s Manual: Hardware", which is not publicly available.
The best I can do is to reply to questions about this PFC module.

> 
> > * We do not support pin group at the moment.
> > * We do not support grandchildren pin definition
> 
> Note that this is a design choice, can't be added later. There's now one driver using grouping model that may be used as a reference in case you are interested: https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/pinctrl/gd%2Cgd32-pinctrl-af.yaml
Thanks for the input!

Commit:
drivers: pinctrl: add Renesas RCar PFC driver

Introduce Pin Function Controller for Renesas R-Car family.
This first implementation support to set a given pin as gpio
or peripheral (GPSR), to set a peripheral function to a
pin (IPSR), to set pull-up, pull-down (PUEN, PUD).
In addition this driver allows to set driving capabilies(DRVCTRL).

Pins are identified thanks to the per SoC binding for
pin definition which also contains pin alternate function
parameters.

Some pins can also have driving capabilities, some have bias
capabilities.
In order to find the correct bias and drive registers pfc_r8a77951.c
describes the different registers.
Each SoCs or package will need to define getters
for these registers: pfc_rcar_get_bias_regs() and
pfc_rcar_get_drive_regs().

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

gmarull commented on include/dt-bindings/pinctrl/renesas/pinctrl-rcar-common.h:
```diff
@@ -0,0 +1,26 @@
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_PINCTRL_RENESAS_PINCTRL_RCAR_COMMON_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_PINCTRL_RENESAS_PINCTRL_RCAR_COMMON_H_
+
+/**
+ * IPSR: Peripheral Function Select Register
+ * Each IPSR bank can hold 8 cellules of 4 bits coded function.
+ *
+ * Function code    [ 0 : 3 ]
+ * Function shift   [ 4 : 9 ]
+ * IPSR bank        [ 10 : 13 ]
+ */
+#define IPSR(bank, shift, func) ((bank << 10) | (shift << 4) | (func))
+
+#define PIN_NOGPSR_START 1024
+/* Pins that can act either as GPIO or peripheral */
+#define RCAR_GP_PIN(bank, pin)     (((bank) * 32) + (pin))
+
+#define RCAR_NOGP_PIN(pin)         (PIN_NOGPSR_START + pin)
```
nit: add complete documentation to public headers, e.g.

```c
/**
 * @brief A brief desc of the macro.
 *
 * @param pin Pin
 */
#define RCAR_NOGP_PIN(pin) (PIN_NOGPSR_START + pin)
```

gmarull commented on dts/arm/renesas/gen3/r8a77951.dtsi:
```diff
@@ -5,3 +5,30 @@
+
+&pfc {
+	can0_data_a_tx: can0_data_a_tx {
+		pins = < PIN_RD FUNC_CAN0_TX_A >;
```
nit: extra spacing `<PIN_RD FUNC_CAN0_TX_A>`

gmarull commented on dts/arm/renesas/gen3/r8a77951.dtsi:
```diff
@@ -5,3 +5,30 @@
+#include <dt-bindings/pinctrl/renesas/pinctrl-r8a77951.h>
+
+&pfc {
+	can0_data_a_tx: can0_data_a_tx {
```
note: pre-defined nodes are, in general, usable for a single state (e.g. `default`). If you plan to add support for other states in the future (e.g. `sleep`), you may want to encode the state name on the node label. Note that it may not always be the case: sleep state for a pin could be represented independently of the mapped signal, e.g. `uart0_tx_pa0`, `spi1_miso_pa0`  for `default` and `pa0_sleep` for `sleep` in any case. See https://docs.zephyrproject.org/latest/guides/pinctrl/index.html#pin-configuration for more details.

gmarull commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -0,0 +1,205 @@
+{
+	const struct pfc_drive_reg *drive_regs =
+		pfc_rcar_get_drive_regs();
+	uint8_t i;
+
+	while (drive_regs->reg) {
+		for (i = 0; i < ARRAY_SIZE(drive_regs->fields); i++) {
```
scope of `i` can be reduced: `for (uint8_t i = 0U...`

gmarull commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -0,0 +1,205 @@
+	uint32_t reg, val;
+
+	reg = pfc_rcar_get_drive_reg(pin, &offset, &size);
+	__ASSERT(reg, "No drive strength register for pin %u", pin);
+
+	if (!reg) {
+		return -EINVAL;
+	}
```
both assert and if at the same time seem unnecessary

gmarull commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -0,0 +1,205 @@
+		return -EINVAL;
+	}
+
+	step = size == 2 ? 24 / 4 : 24 / 8;
```
some explained macros would help understanding this better

gmarull commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -0,0 +1,205 @@
+							      uint8_t *bit)
+{
+	const struct pfc_bias_reg *bias_regs = pfc_rcar_get_bias_regs();
+	size_t i;
+
+	/* Loop arround all the registers to find the bit for a given pin */
+	while (bias_regs->puen && bias_regs->pud) {
+		for (i = 0; i < ARRAY_SIZE(bias_regs->pins); i++) {
```
ditto

gmarull commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -0,0 +1,205 @@
+}
+
+static const struct pfc_bias_reg *pfc_rcar_get_bias_reg(uint16_t pin,
+							      uint8_t *bit)
```
nit: align second parameter properly

gmarull commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -0,0 +1,205 @@
+}
+
+
+static uint32_t pfc_rcar_get_drive_reg(uint16_t pin, uint8_t *offset,
+					  uint8_t *size)
```
nit: align second line properly

gmarull commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -0,0 +1,205 @@
+	const struct pfc_bias_reg *bias_reg =
+		pfc_rcar_get_bias_reg(pin, &bit);
+
+	__ASSERT(bias_reg, "No bias register for pin %u", pin);
+	if (!bias_reg) {
+		return -EINVAL;
+	}
```
ditto

gmarull commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -0,0 +1,205 @@
+{
+	uint32_t val;
+	uint8_t bit;
+	const struct pfc_bias_reg *bias_reg =
+		pfc_rcar_get_bias_reg(pin, &bit);
```
fits into 80cols?

gmarull commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -0,0 +1,205 @@
+
+	if (pin->flags & RCAR_PIN_FLAGS_PULL_SET) {
+		ret = pfc_rcar_set_bias(pin->pin, pin->flags);
+		if (ret) {
```
this and otthers: better use essentially boolean expressions: `if (ret < 0)`

gmarull commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -0,0 +1,205 @@
+}
+
+int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt,
+			   uintptr_t reg)
```
`ARG_UNUSED(reg)`

gmarull commented on soc/arm/renesas_rcar/gen3/Kconfig.series:
```diff
@@ -10,5 +10,6 @@ config SOC_SERIES_RCAR_GEN3
 	select CPU_HAS_DCLS
 	select SOC_FAMILY_RCAR
 	select CLOCK_CONTROL_RCAR_CPG_MSSR if CLOCK_CONTROL
+	select PINCTRL
```
I'd suggest instead this in `Kconfig.defconfig`

```c
config PINCTRL
    default y
```

(doesn't make it a strong requirement)

gmarull commented on soc/arm/renesas_rcar/gen3/pfc_r8a77951.c:
```diff
@@ -0,0 +1,517 @@
+	{ PFC_BIAS_REG("PUEN0", 0x0400, "PUD0", 0x0440) {
+		  [0]  = PIN_QSPI0_SPCLK,       /* QSPI0_SPCLK */
+		  [1]  = PIN_QSPI0_MOSI_IO0,    /* QSPI0_MOSI_IO0 */
+		  [2]  = PIN_QSPI0_MISO_IO1,    /* QSPI0_MISO_IO1 */
+		  [3]  = PIN_QSPI0_IO2,	       /* QSPI0_IO2 */
+		  [4]  = PIN_QSPI0_IO3,	       /* QSPI0_IO3 */
+		  [5]  = PIN_QSPI0_SSL,	       /* QSPI0_SSL */
+		  [6]  = PIN_QSPI1_SPCLK,       /* QSPI1_SPCLK */
+		  [7]  = PIN_QSPI1_MOSI_IO0,    /* QSPI1_MOSI_IO0 */
+		  [8]  = PIN_QSPI1_MISO_IO1,    /* QSPI1_MISO_IO1 */
+		  [9]  =  PIN_QSPI1_IO2,	       /* QSPI1_IO2 */
+		  [10] = PIN_QSPI1_IO3,         /* QSPI1_IO3 */
```
nit: alignment here

gmarull commented on soc/arm/renesas_rcar/gen3/pinctrl_soc.h:
```diff
@@ -0,0 +1,111 @@
+
+#define RCAR_DT_PIN(node_id)							    \
+	{									    \
+		.pin = DT_PROP_BY_IDX(node_id, pins, 0),			    \
+		.func =								    \
+			COND_CODE_1(DT_PROP_HAS_IDX(node_id, pins, 1),		    \
+				    (RCAR_PIN_FUNC(node_id)), (0)),		    \
+		.flags = RCAR_PIN_FLAGS(node_id),				    \
+		.drive_strength =						    \
+			COND_CODE_1(DT_NODE_HAS_PROP(node_id, drive_strength),	    \
+				    (DT_PROP(node_id, drive_strength)), (0)),	    \
```
```c
		.pin = DT_PROP_BY_IDX(node_id, pins, 0),		       \
		.func = COND_CODE_1(DT_PROP_HAS_IDX(node_id, pins, 1),	       \
				    (RCAR_PIN_FUNC(node_id)), (0)),	       \
		.flags = RCAR_PIN_FLAGS(node_id),			       \
		.drive_strength =					       \
			COND_CODE_1(DT_NODE_HAS_PROP(node_id, drive_strength), \
				    (DT_PROP(node_id, drive_strength)), (0)),  \
```

gmarull commented on soc/arm/renesas_rcar/gen3/pinctrl_soc.h:
```diff
@@ -0,0 +1,111 @@
+const struct pfc_bias_reg *pfc_rcar_get_bias_regs(void);
+const struct pfc_drive_reg *pfc_rcar_get_drive_regs(void);
+
+#define PIN_NOGPSR_START 1024
```
isn't this already defined?

gmarull commented on soc/arm/renesas_rcar/gen3/pinctrl_soc.h:
```diff
@@ -0,0 +1,111 @@
+};
+
+struct pfc_bias_reg {
+	uint32_t puen;		/* Pull-enable or pull-up control register */
+	uint32_t pud;		/* Pull-up/down or pull-down control register */
+	const uint16_t pins[32];
```
please document using Doxygen format `/**`

gmarull commented on soc/arm/renesas_rcar/gen3/pinctrl_soc.h:
```diff
@@ -0,0 +1,111 @@
+	DT_PROP(node_id, bias_pull_up)   * RCAR_PIN_PULL_UP |		       \
+	DT_PROP(node_id, bias_pull_down) * RCAR_PIN_PULL_DOWN |		       \
+	DT_PROP(node_id, bias_disable)   * RCAR_PIN_PULL_DISABLE |	       \
+	DT_NODE_HAS_PROP(node_id, drive_strength) * RCAR_PIN_FLAGS_DRIVE_SET | \
```
suggestion (ignore if not accurate): `drive_strength == 0` could indicate this

gmarull replied:
Looking at line 64: looks like drive_strength == 0 is equivalent to !DT_NODE_HAS_PROP(node_id, drive_strength)

gmarull commented on soc/arm/renesas_rcar/gen3/pinctrl_soc.h:
```diff
@@ -0,0 +1,111 @@
+/* Offsets are defined in dt-bindings pinctrl-rcar-common.h */
+#define RCAR_PIN_FUNC(node_id)			     \
+	{					     \
+		((RCAR_IPSR(node_id) >> 10) & 0x1F), \
+		((RCAR_IPSR(node_id) >> 4) & 0x7),   \
+		((RCAR_IPSR(node_id) & 0xF))	     \
```
nit: add `U` suffix to unsigned constants here

gmarull commented on soc/arm/renesas_rcar/gen3/pinctrl_soc.h:
```diff
@@ -0,0 +1,111 @@
+	uint8_t drive_strength;
+} pinctrl_soc_pin_t;
+
+#define RCAR_IPSR(node_id) \
+	DT_PROP_BY_IDX(node_id, pins, 1)
```
`#define RCAR_HAS_IPSR(node_id) DT_PROP_HAS_IDX(node_id, pins, 1)` can likely be added (for improved code readability)

gmarull commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -0,0 +1,204 @@
+	uint32_t reg, val;
+
+	reg = pfc_rcar_get_drive_reg(pin, &offset, &size);
+	if (reg == 0) {
```
```suggestion
	if (reg == 0U) {
```

gmarull commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -0,0 +1,204 @@
+	}
+
+	step = PFC_RCAR_DRIVE_STEP(size);
+	if (strength < step || strength > 24) {
```
```suggestion
	if ((strength < step) || (strength > 24U)) {
```

gmarull commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -0,0 +1,204 @@
+	/* Convert the value from mA based on a full drive strength
+	 * value of 24mA.
+	 */
+	strength = strength / step - 1;
```
```suggestion
	strength = (strength / step) - 1U;
```

gmarull commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -0,0 +1,204 @@
+	strength = strength / step - 1;
+	/* clear previous drive strength value */
+	val = sys_read32(PFC_REG_BASE + reg);
+	val &= ~GENMASK(offset + size - 1, offset);
```
```suggestion
	val &= ~GENMASK(offset + size - 1U, offset);
```

gmarull commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -0,0 +1,204 @@
+
+	/* Loop arround all the registers to find the bit for a given pin */
+	while (bias_regs->puen && bias_regs->pud) {
+		for (size_t i = 0; i < ARRAY_SIZE(bias_regs->pins); i++) {
```
```suggestion
		for (size_t i = 0U; i < ARRAY_SIZE(bias_regs->pins); i++) {
```

gmarull commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -0,0 +1,204 @@
+
+	/* pull enable/disable*/
+	val = sys_read32(PFC_REG_BASE + bias_reg->puen);
+	if ((flags & RCAR_PIN_FLAGS_PUEN) == 0) {
```
```suggestion
	if ((flags & RCAR_PIN_FLAGS_PUEN) == 0U) {
```

gmarull commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -0,0 +1,204 @@
+		pfc_rcar_set_gpsr(pin->pin, true);
+	}
+
+	if (pin->flags & RCAR_PIN_FLAGS_FUNC_SET) {
```
```suggestion
	if ((pin->flags & RCAR_PIN_FLAGS_FUNC_SET) != 0U) {
```

gmarull commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -0,0 +1,204 @@
+		pfc_rcar_set_ipsr(&pin->func);
+	}
+
+	if (pin->flags & RCAR_PIN_FLAGS_PULL_SET) {
```
```suggestion
	if ((pin->flags & RCAR_PIN_FLAGS_PULL_SET) != 0U) {
```

gmarull commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -0,0 +1,204 @@
+	int ret = 0;
+
+	ARG_UNUSED(reg);
+	while (pin_cnt--) {
```
```suggestion
	while (pin_cnt-- > 0U) {
```

gmarull commented on dts/bindings/pinctrl/renesas,rcar-pfc.yaml:
```diff
@@ -0,0 +1,33 @@
+# Copyright (c) 2021 IoT.bzh
+# SPDX-License-Identifier: Apache-2.0
+
+description: Renesas R-Car Pin Function Controller node
```
consider adding more detailed documentation, see for example nrf or gd32 bindings

gmarull commented on include/dt-bindings/pinctrl/renesas/pinctrl-rcar-common.h:
```diff
@@ -0,0 +1,43 @@
+ * Function shift   [ 4 : 9 ]
+ * IPSR bank        [ 10 : 13 ]
+ */
+#define IPSR(bank, shift, func) ((bank << 10) | (shift << 4) | (func))
```
```suggestion
#define IPSR(bank, shift, func) (((bank) << 10U) | ((shift) << 4U) | (func))
```

gmarull commented on include/dt-bindings/pinctrl/renesas/pinctrl-rcar-common.h:
```diff
@@ -0,0 +1,43 @@
+ * @param bank the GPIO bank
+ * @param pin the pin within the GPIO bank (0..31)
+ */
+#define RCAR_GP_PIN(bank, pin)     (((bank) * 32) + (pin))
```
```suggestion
#define RCAR_GP_PIN(bank, pin)     (((bank) * 32U) + (pin))
```

gmarull commented on include/dt-bindings/pinctrl/renesas/pinctrl-rcar-common.h:
```diff
@@ -0,0 +1,43 @@
+#define IPSR(bank, shift, func) ((bank << 10) | (shift << 4) | (func))
+
+/* Arbitrary number to encode non capable gpio pin */
+#define PIN_NOGPSR_START 1024
```
```suggestion
#define PIN_NOGPSR_START 1024U
```

gmarull commented on soc/arm/renesas_rcar/gen3/pfc_r8a77951.c:
```diff
@@ -0,0 +1,517 @@
+#include "pinctrl_soc.h"
+#include <dt-bindings/pinctrl/renesas/pinctrl-r8a77951.h>
+
+#define PFC_DRIVE_REG(name, r) \
+	.reg = r,		  \
+	.fields =
```
`name` is not used by the macro?

gmarull commented on soc/arm/renesas_rcar/gen3/pfc_r8a77951.c:
```diff
@@ -0,0 +1,517 @@
+	{ },
+};
+
+#define PFC_BIAS_REG(name1, r1, name2, r2) \
+	.puen = r1,			   \
+	.pud = r2,			   \
+	.pins =
```
ditto, name1/2 here. If these are intended to be used as a comment use `/* */`

gmarull commented on soc/arm/renesas_rcar/gen3/pinctrl_soc.h:
```diff
@@ -0,0 +1,110 @@
+#include <stdbool.h>
+#include <sys/util.h>
+
+struct rcar_pin_func {
+	uint8_t bank:5;      /* bank number 0 - 18 */
+	uint8_t shift:5;     /* bit shift 0 - 28 */
+	uint8_t func:4;      /* choice from 0x0 to 0xF */
+};
+
+#define RCAR_PIN_FLAGS_PULL_SET       BIT(0)
+#define RCAR_PIN_FLAGS_PUEN           BIT(1)
+#define RCAR_PIN_FLAGS_PUD            BIT(2)
+#define RCAR_PIN_FLAGS_DRIVE_SET      BIT(3)
+#define RCAR_PIN_FLAGS_FUNC_SET       BIT(4)
+
+#define RCAR_PIN_PULL_UP      (RCAR_PIN_FLAGS_PULL_SET | RCAR_PIN_FLAGS_PUEN | RCAR_PIN_FLAGS_PUD)
+#define RCAR_PIN_PULL_DOWN    (RCAR_PIN_FLAGS_PULL_SET | RCAR_PIN_FLAGS_PUEN | RCAR_PIN_FLAGS_PUD)
+#define RCAR_PIN_PULL_DISABLE  RCAR_PIN_FLAGS_PULL_SET
+
+/** Type for R-Car pin. */
+typedef struct pinctrl_soc_pin {
+	uint16_t pin;
+	struct rcar_pin_func func;
+	uint8_t flags;
+	uint8_t drive_strength;
+} pinctrl_soc_pin_t;
+
+#define RCAR_IPSR(node_id) DT_PROP_BY_IDX(node_id, pins, 1)
+#define RCAR_HAS_IPSR(node_id) DT_PROP_HAS_IDX(node_id, pins, 1)
+
+/* Offsets are defined in dt-bindings pinctrl-rcar-common.h */
+#define RCAR_PIN_FUNC(node_id)			       \
+	{					       \
+		((RCAR_IPSR(node_id) >> 10U) & 0x1FU), \
+		((RCAR_IPSR(node_id) >> 4U) & 0x7U),   \
+		((RCAR_IPSR(node_id) & 0xFU))	       \
+	}
+
+#define RCAR_PIN_FLAGS(node_id)						       \
```
I'd suggest to document a little bit some of these macros/values.

gmarull reviewed (commented):
I have some minor comments/suggestions (ignore some if not meaningful), in general LGTM

julien-massot commented:
HI @gmarull just finished to address all your suggestion.
Thank you so much for all of them !
I also improved the binding documentation as suggested.

gmarull commented on dts/arm/renesas/gen3/r8a77951.dtsi:
```diff
@@ -5,3 +5,30 @@
+
+&pfc {
+	can0_data_a_tx_default: can0_data_a_tx_default {
+		pins = <PIN_RD FUNC_CAN0_TX_A>;
```
question: does each entry define more than one pin? (`pins` is plural)

julien-massot replied:
Only one, will turn it into 'pin'.

gmarull reviewed (commented):
Overall looks good, I have a few questions though.

julien-massot commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -0,0 +1,204 @@
+ *
+ */
+
+#include <devicetree.h>
+#include <drivers/pinctrl.h>
+#include <soc.h>
+#include <sys/util.h>
+
+#define DT_DRV_COMPAT renesas_rcar_pfc
```
ok

julien-massot commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -0,0 +1,204 @@
+	}
+
+	step = PFC_RCAR_DRIVE_STEP(size);
+	if ((strength < step) || (strength > 24U)) {
```
Thanks, will do

julien-massot commented on dts/arm/renesas/gen3/r8a77951.dtsi:
```diff
@@ -5,3 +5,30 @@
+
+	scif2_data_a_tx_default: scif2_data_a_tx_default {
+		pins = <PIN_TX2_A FUNC_TX2_A>;
+	};
+	scif2_data_a_rx_default: scif2_data_a_rx_default {
+
+		pins = <PIN_RX2_A FUNC_RX2_A>;
```
thanks

julien-massot commented on dts/bindings/pinctrl/renesas,rcar-pfc.yaml:
```diff
@@ -0,0 +1,93 @@
+
+    All device pin configurations should be placed in child nodes of the
+    'pfc' node, as shown in this example:
+      /* You can put this in places like a board-pinctrl.dtsi file in
+       * your board directory, or a devicetree overlay in your application.
+       */
```
Well indeed these pre-defined node, should not be done in the SoC dtsi, that's fishy.

That's up to to the board dts to decide if FUNC_CAN0_TX_A should be routed to PIN_RD,
or to use this pin for another function.

gmarull replied:
I think that:

1) If there's a limited valid set (e.g. case of STM32, GD32, NXP...) pre-definitions are a good idea, but I think they should be provided independently of the supported boards upstream (i.e. all of them). This makes sure that the usage is the same for both, upstream and downstream.
2) If there's _complete freedom_ (e.g. Nordic), then pre-definitions don't make sense, all boards need to define the combination they choose.

julien-massot replied:
just pushed a new version but will do a new one with your suggestion '2'.

There is so much possibility on R-Car that it doesn't make sense to declare all of them.
Some possibilities will be really unlikely to be used in a Zephyr context.

Thanks

julien-massot commented on soc/arm/renesas_rcar/gen3/pinctrl_soc.h:
```diff
@@ -0,0 +1,117 @@
+ *
+ */
+
+#ifndef ZEPHYR_SOC_ARM_RENESAS_RCAR_GEN3_PINCTRL_SOC_H_
+#define ZEPHYR_SOC_ARM_RENESAS_RCAR_GEN3_PINCTRL_SOC_H_
+#include <devicetree.h>
+#include <dt-bindings/pinctrl/renesas/pinctrl-rcar-common.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <sys/util.h>
```
ok

Commit:
boards: arm: rcar_h3ulcb: add pins definition

Add some pins definitions for CAN0 and SCIF1/2.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>

gmarull reviewed (approved):
LGTM, one minor comment

julien-massot commented on boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7.dts:
```diff
@@ -52,6 +52,8 @@
 };
 
 &can0 {
+	pinctrl-0 = <&can0_data_a_tx_default &can0_data_a_rx_default>;
+	pinctrl-names = "default";
```
just moved to the board-pinctrl commit

carlescufi commented:
Compliance failure is expected, new pin typedef.
